spatialdata/_core/models.py

```diff
@@ -0,0 +1,421 @@
+"""This file contains models and schema for SpatialData"""
+
+from functools import singledispatch
+from typing import Any, Dict, Literal, Mapping, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from dask.array.core import Array as DaskArray
+from dask.array.core import from_array
+from geopandas import GeoDataFrame
+from geopandas.array import GeometryDtype
+from multiscale_spatial_image import to_multiscale
+from multiscale_spatial_image.multiscale_spatial_image import MultiscaleSpatialImage
+from multiscale_spatial_image.to_multiscale.to_multiscale import Methods
+from numpy.typing import ArrayLike
+from pandas.api.types import is_categorical_dtype
+from pandera import Column, DataFrameSchema
+from shapely import GeometryType
+from shapely.io import from_geojson, from_ragged_array
+from spatial_image import SpatialImage, to_spatial_image
+from xarray_schema.components import ArrayTypeSchema, AttrSchema, DimsSchema
+from xarray_schema.dataarray import DataArraySchema
+
+from spatialdata._constants._constants import RasterType
+from spatialdata._core.transformations import BaseTransformation, Identity
+from spatialdata._logging import logger
+
+# Types
+Chunks_t = Union[
+    int,
+    Tuple[int, ...],
+    Tuple[Tuple[int, ...], ...],
+    Mapping[Any, Union[None, int, Tuple[int, ...]]],
+]
+ScaleFactors_t = Sequence[Union[Dict[str, int], int]]
+
+Z, C, Y, X = "z", "c", "y", "x"
+
+Transform_s = AttrSchema(BaseTransformation, None)
+
+Labels2D_s = DataArraySchema(
+    dims=DimsSchema((Y, X)),
+    array_type=ArrayTypeSchema(DaskArray),
+    attrs={"transform": Transform_s},
+)
+Image2D_s = DataArraySchema(
+    dims=DimsSchema((C, Y, X)),
+    array_type=ArrayTypeSchema(DaskArray),
+    attrs={"transform": Transform_s},
+)
+Labels3D_s = DataArraySchema(
+    dims=DimsSchema((Z, Y, X)),
+    array_type=ArrayTypeSchema(DaskArray),
+    attrs={"transform": Transform_s},
+)
+Image3D_s = DataArraySchema(
+    dims=DimsSchema((Z, C, Y, X)),
```

Comment:
Let's use czyx to reflect the NGFF 0.4 specs. Will be nice to support arbitrary order of axes but this maybe has to be done on the `SpatialImage` side.

Reply:
done